### PR TITLE
9.x/reducer

### PR DESCRIPTION
**Pull request by pmccormick**
Billy's reducer implementation.  Indentation was messy...  While not perfect hopefully this is easier to merge in...   Note this was originally based on 8.x and to get it build I had to hack some things (e.g., `Function*` became `FuntionCallee`).  

Pushing this before a lot of testing as I don't really have anything to do any significant testing with yet but don't want to drift too far off the code base.

### CODE REVIEW

**Comment by pmccormick** [comment]
This should have gone against 9.x -- not sure where it reverted to 8.x on me...  Perhaps because we don't have 9.x as our default?

File: llvm/reducertest/intsummod.ll [hunk 1]
```diff
@@ -0,0 +1,430 @@
+ivate unnamed_addr constant [26 x i8] c"Result: %d/%d successes!\0A\00", align 1
```
Comment (stelleg) [inline llvm/reducertest/intsummod.ll]: Looks like the example got cut off.

File: clang/lib/CodeGen/CGExpr.cpp [hunk 2]
```diff
@@ -2576,20 +2576,20 @@ LValue CodeGenFunction::EmitDeclRefLValue(const DeclRefExpr *E) {
     // Check for captured variables.
     if (E->refersToEnclosingVariableOrCapture()) {
 
-      // kitsune: if we are generating a kokkos-based lambda construct
```
Comment (stelleg) [inline clang/lib/CodeGen/CGExpr.cpp]: Looks like all the changes in this file are unrelated to reducer.

**Comment by pmccormick** [comment]
> Overall looks pretty good to me, except for the commented parts. I also think it would be good to have a C example (or preferably test) of how it is intended to be used, because it's not particularly obvious from the code itself.

Billy is working on one.  👍

File: llvm/reducertest/intsummod.ll (hunk 1 above)
Comment (pmccormick) [reply llvm/reducertest/intsummod.ll]: Interesting...  It should have done a straight copy -- I'll double check it again...

File: clang/lib/CodeGen/CGExpr.cpp (hunk 2 above)
Comment (pmccormick) [reply clang/lib/CodeGen/CGExpr.cpp]: Yeah, more changes coming related to this path -- I thought I stashed them for this PR but this  slipped through.  I'm tempted to leave as-is for now since it will just be part of the next PR anyways and this is minimal...

**Comment by pmccormick** [comment]
Shelving this as it is way too far out of date now... More work needs to be done here...
